src/quartics/ferrari_qs_functor.py

```diff
@@ -0,0 +1,216 @@
+
+A Ferrari/Cardano solver implementing mpmath to allow for arbitrary precision
+
+Note
+----
+Created with reference to https://github.com/NKrvavica/fqs, whose MIT License statement,
+should it be necessary, is included at the end of the file.
+"""
+
+import mpmath
+from mpmath import mpf, mpc
+from math import isclose
+
+
+MpfAble: type = float|int|str|mpf
+
+
+class SolveFerrari:
+    def __init__(self, coeffs: list[MpfAble]):
+        '''
+        Returns
+        -------
+        A list of (potentially complex) roots of the given quartic polynomial, as determined by the Ferrari-Cardano method 
+
+        Parameters
+        ----------
+        coeffs : list[MpfAble]
+            The coefficients of the quartic polynomial to solve, c[0]x^4 + c[1]x^3 + c[2]x^3 + c[3]x + c[4]
+        '''
+        if not all(isinstance(c, MpfAble) for c in coeffs):
+            raise ValueError("All coefficients must be either mpf instances, or float, int, str which can be converted thereinto")
+        if not len(coeffs) in [4,5]:
+            raise ValueError("Coefficients must either be a full 5 for a quartic, or the last 4 of one that has already been normalized (a=1)")
+        
+        coeffs = [mpf(c) for c in coeffs]
+
+        if len(coeffs) == 5:
+            print("Need to normalize!")
+            a = coeffs[0]
+            coeffs = [coeffs[i]/a for i in range(len(coeffs))]
+        else:
+            coeffs.insert(0,mpf(1))
+
+        self._coeffs: list[mpf] = coeffs
+
+    def __call__(self, *args, **kwds) -> list[mpc]:
+        
+        return self._solve_normalized_quartic()
+
+    def _solve_normalized_quadratic(self, b, c) -> tuple[mpc]:
+        '''
+        Returns
+        -------
+        The (potentially complex) roots of the given quadratic
+
+        Parameters
+        ----------
+        b, c : mpf
+            The coefficients of the quadratic x^2 + bx + c = 0
+        '''
+        assert type(b) == type(c) == mpf
+
+        # Predivide by 2
+        b0 = -0.5*b
+        discrim = b0*b0 - c
+        sqrt_discrim = mpmath.sqrt(discrim)
+
+        # Roots
+        r1 = b0 - sqrt_discrim
+        r2 = b0 + sqrt_discrim
+
+        return r1, r2
+    
+    def _one_real_root_of_normalized_cubic(self, b, c, d) -> mpf:
+        '''
+        Returns
+        -------
+        A real root for the given cubic equation, since at least one of the three must be real.
+
+        Parameters
+        ----------
+        b, c, d : mpf
+            The coeffients of the normalized cubic equation x^3 + bx^2 + cx + d = 0
+        '''
+        assert type(b) == type(c) == type(d) == mpf
+
+        # Repeating values
+        third = mpf(1)/mpf(3)
+        third_b = b*third
+        third_b2 = sq(third_b)
+
+        # Intermediate variables
+        f = third*c - third_b2
+        g = third_b * (2*third_b2 - c) + d
+        h = 0.25*sq(g) + f**3
+
+        if f == g == h == mpf(0): #Should this be converted to closeness test?
```

Comment:
Probably! Make a helper function to check closeness using the 2^prec logic like you do below.